packages/type-system/rust/src/ontology/uri/native.rs

```diff
@@ -0,0 +1,34 @@
+use url::Url;
+
+use crate::ontology::uri::{error::ParseBaseUriError, BaseUri, VersionedUri};
+
+impl BaseUri {
+    pub fn new(uri: &str) -> Result<BaseUri, ParseBaseUriError> {
+        // TODO: Propagate more useful errors
+        // TODO: This attempts to parse the string _into_ a valid URL. Perhaps we want to enforce
+        //  that the string is valid (by checking the output is equal to the input). An example:
+        //  "file://loc%61lhost/" is turned into "file:///"
```

Comment:
This is quite strange behaviour but seems somewhat consistent with the reference implementation of the actual [URL spec](https://url.spec.whatwg.org/). I'm not sure what we want to do here because it does helpfully remove stray slashes and standardises some URLs, but at the same time the example I gave here seems like something we'd probably want errors for (things that _are_ considered errors by the library are [here](https://docs.rs/url/latest/url/enum.ParseError.html)).